### ISSUE TITLE
Update CTE support for various providers

### ISSUE THREAD

**Issue by viceroypenguin** [linq2db/linq2db#3015]
Made some tests with failing providers:
- SAP HANA: supports CTE in source query (CTE should be moved inside source clause as part of source query) but linq2db doesn't support CTE for SAP HANA right now
- Oracle: also supports cte if it is a part of source query. Also oracle doesn't like `_` as table alias
- DB2: version 11.5 we use on CI declares CTE support in target so it makes sense to test it (note that syntax is differ from sql server: https://www.ibm.com/docs/en/db2/11.5?topic=statements-merge). Also it makes sense to test if source CTE as part of select will work
- informix: CTE support added in v14 we use on CI, makes sense to test is as part of source query
- firebird: CTE works in source if moved to source query

so basically:
- for most of providers we should move source CTE to source query ~when we generate SQL~ actually I think it should be moved to source query in AST
- test CTE in target for DB2
- unblock CTE support for SAP HANA and enable CTE tests for it

_Originally posted by @MaceWindu in https://github.com/linq2db/linq2db/issues/2363#issuecomment-846415519_
